File: backend/RAG/src/embeddings.py
import os
import time
import logging
from openai import AzureOpenAI
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)


class EmbeddingManager:
    def __init__(self):
        self.api_key = os.getenv("AZURE_OPENAI_API_KEY")
        self.api_endpoint = os.getenv("AZURE_OPENAI_ENDPOINT")
        self.api_version = os.getenv("AZURE_OPENAI_API_VERSION")
        self.deployment_name = os.getenv("AZURE_OPENAI_EMBEDDING_DEPLOYMENT")
        self.client = None

    def initialize(self):
        """Initialize Azure OpenAI embedding model."""
        try:
            self.client = AzureOpenAI(
                api_key=self.api_key,
                api_version=self.api_version,
                azure_endpoint=self.api_endpoint
            )
            logger.info(f"Embedding model '{self.deployment_name}' initialized.")
            return True
        except Exception as e:
            logger.error(f"Error initializing embedding model: {e}")
            return False

    def generate_embedding(self, text):
        """Generate embedding for a single text."""
        try:
            start_time = time.time()
            text = text.replace("\n", " ").strip()
            response = self.client.embeddings.create(
                input=text,
                model=self.deployment_name
            )
            elapsed = time.time() - start_time
            logger.info(f"[TIMING] Single embedding generated in {elapsed:.2f}s")
            return response.data[0].embedding
        except Exception as e:
            logger.error(f"Error generating embedding: {e}")
            return None

    def batch_generate_embeddings(self, texts):
        """Generate embeddings for multiple texts."""
        try:
            start_time = time.time()
            response = self.client.embeddings.create(
                input=texts,
                model=self.deployment_name
            )
            embeddings = [item.embedding for item in response.data]
            elapsed = time.time() - start_time
            logger.info(f"[TIMING] Batch embeddings ({len(texts)} texts) generated in {elapsed:.2f}s")
            return embeddings
        except Exception as e:
            logger.error(f"Error batch generating embeddings: {e}")
            return None

File: senacor-voicelive/lib/rag/embeddings.ts
/**
 * Embedding Manager
 * Migrated from: backend/RAG/src/embeddings.py
 *
 * Azure OpenAI embedding generation for RAG pipeline.
 */

import { AzureOpenAI } from 'openai'

export class EmbeddingManager {
  private apiKey: string
  private apiEndpoint: string
  private apiVersion: string
  private deploymentName: string
  private client: AzureOpenAI | null = null

  constructor() {
    this.apiKey = process.env.AZURE_OPENAI_API_KEY ?? ''
    this.apiEndpoint = process.env.AZURE_OPENAI_ENDPOINT ?? ''
    this.apiVersion = process.env.AZURE_OPENAI_API_VERSION ?? ''
    this.deploymentName = process.env.AZURE_OPENAI_EMBEDDING_DEPLOYMENT ?? ''
  }

  /** Initialize Azure OpenAI embedding model. */
  initialize(): boolean {
    try {
      this.client = new AzureOpenAI({
        apiKey: this.apiKey,
        apiVersion: this.apiVersion,
        endpoint: this.apiEndpoint,
      })
      console.log(`Embedding model '${this.deploymentName}' initialized.`)
      return true
    } catch (e) {
      console.error(`Error initializing embedding model: ${e}`)
      return false
    }
  }

  /** Generate embedding for a single text. */
  async generateEmbedding(text: string): Promise<number[] | null> {
    try {
      if (!this.client) throw new Error('Not initialized')

      const startTime = performance.now()
      const cleanText = text.replace(/\n/g, ' ').trim()

      const response = await this.client.embeddings.create({
        input: cleanText,
        model: this.deploymentName,
      })

      const elapsed = (performance.now() - startTime) / 1000
      console.log(`[TIMING] Single embedding generated in ${elapsed.toFixed(2)}s`)

      return response.data[0].embedding
    } catch (e) {
      console.error(`Error generating embedding: ${e}`)
      return null
    }
  }

  /** Generate embeddings for multiple texts. */
  async batchGenerateEmbeddings(texts: string[]): Promise<number[][] | null> {
    try {
      if (!this.client) throw new Error('Not initialized')

      const startTime = performance.now()

      const response = await this.client.embeddings.create({
        input: texts,
        model: this.deploymentName,
      })

      const embeddings = response.data.map((item) => item.embedding)
      const elapsed = (performance.now() - startTime) / 1000
      console.log(`[TIMING] Batch embeddings (${texts.length} texts) generated in ${elapsed.toFixed(2)}s`)

      return embeddings
    } catch (e) {
      console.error(`Error batch generating embeddings: ${e}`)
      return null
    }
  }
}
